homeassistant/components/utility_meter/__init__.py

```diff
@@ -275,3 +266,28 @@ async def async_migrate_entry(hass: HomeAssistant, config_entry: ConfigEntry) ->
     _LOGGER.info("Migration to version %s successful", config_entry.version)
 
     return True
+
+
+async def async_remove_obsolete_devices_entry(
+    hass: HomeAssistant, entry_id: str, entity_id: str
+) -> None:
+    """Remove device link for entry, the source device may have changed."""
+
+    device_registry = dr.async_get(hass)
+    entity_registry = er.async_get(hass)
+
+    # Resolve source entity device
+    current_device_id = None
+    if ((source_entity := entity_registry.async_get(entity_id)) is not None) and (
+        source_entity.device_id is not None
+    ):
+        current_device_id = source_entity.device_id
+
+    devices_in_entry = device_registry.devices.get_devices_for_config_entry_id(entry_id)
+
+    # Removes all devices from the config entry that are not the same as the current device
+    for device_in_entry in devices_in_entry:
+        if device_in_entry.id != current_device_id:
+            device_registry.async_update_device(
+                device_in_entry.id, remove_config_entry_id=entry_id
+            )
```

Comment:
```suggestion
    for device in devices_in_entry:
        if device.id == current_device_id:
            continue
        device_registry.async_update_device(device.id, remove_config_entry_id=entry_id)
```

Reply:
Wont work. The device is not guaranteed to be found.

Reply:
And there could be many

Reply:
It seems to work well with 3 devices...

Reply:
It will only ever remove the link from one device. The function is only called once.

So if, for some reason, it ends up linked to two devices by mistake or whatever, this wont clean it up.

Reply:
Så say a entity id is renamed, causing it to move between two devices. I dont think we actually trigger re association  right now. So it probably wont cause double links at the moment.

But we already have users now with double links, which wont be cleaned up by this unless they move the helper back to each one that is linked and away again.

Reply:
> It will only ever remove the link from one device. The function is only called once.
> 
> So if, for some reason, it ends up linked to two devices by mistake or whatever, this wont clean it up.

I disagree, the for loop is still present, if the device id is equal to the correct device it will go to the next loop without deleting the link. The test was also changed to a three-device case (one correct and two stale):
https://github.com/home-assistant/core/blob/8321835ad35270eae0bc6089c3f99f1bdf590009/homeassistant/components/utility_meter/__init__.py#L289-L292

I understand that this resolves the users' double links, as this cleaning is in the setup step of the config entry, which was one of the reasons for moving the cleaning to the setup entry.

Reply:
The only device that get the async_update_device call is the last device in the for loop ie the value device holds when the for loop exists.

for x in [a,b,c]
 pass
x == c

for x in [a]
 if x != a continue
x == a

So if the list only contain the current device it will remove that anyway.

Reply:
I couldn't see what you pointed out...

The current test contains three devices that work, I did local tests with debug log and everything seems to work properly:

* Just the correct device:
```
Loop 1: Device 2f2ce9444b2ad1f1696769d6035607e2 kept in config entry
```

* With three devices, one correct and two stales:
```
Loop 1: Device 5c0175c6d79c6d272cc0190cd8cb43be removed from config entry
Loop 2: Device 2f2ce9444b2ad1f1696769d6035607e2 kept in config entry
Loop 3: Device c97f157fe6f1486da02aef681472c627 removed from config entry
```

Please, can you suggest a change that addresses what you point out?

Reply:
What the heck. I've looked at this on a phone and it was showing a different indentation level of the remove call than what i see now. Now I see that the removal is inside the loop.

Sorry if this was the case all the time. This should work...